src/libopensc/card-piv.c

```diff
@@ -2715,33 +2720,32 @@ static int piv_process_discovery(sc_card_t *card)
 
 	priv->ccc_flags |= PIV_CCC_FOUND;
 
-	/* CCC boby has 1 byte tags, and 1 byte lengths */
-	p = body;
-	end = p + bodylen;
-	while (p <= (end - 2) && (next = (p + *(p + 1) + 2)) <= end ) {
-		switch (*p) {
-			case 0xf0:
-				if (*(p + 1) == 0x15) {
-					if (memcmp((p + 2) ,"\xA0\x00\x00\x03\08", 5) == 0)
+	/* CCC  entries are simple tlv */
+	end = body + bodylen;
+	
+	for(; (body < end); body += len) {
+	
+		r = sc_simpletlv_read_tag((u8**)&body, end - body , &tag, &len);
+		if (r < 0)
+			goto err;
+		switch (tag) {
+			case PIV_CCC_TAG_F0:
```

Comment:
nit: Switch and case can have the same indent to avoid too much indentation. I would recommend it in this case.